File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/presentation/ui/base/activity/BaseAuthenticatedActivity.java
package com.lattis.ellipse.presentation.ui.base.activity;

import android.accounts.Account;
import android.accounts.AccountManager;
import androidx.annotation.Nullable;

import io.lattis.ellipse.R;


public abstract class BaseAuthenticatedActivity<Presenter extends ActivityPresenter> extends BaseActivity<Presenter> {

    @Override
    protected void onResume() {
        authenticateAccount();
        super.onResume();
    }

    @SuppressWarnings("MissingPermission")
    private boolean authenticateAccount() {
        Account account = getAccount();
        AccountManager accountManager = AccountManager.get(this);
        if (account != null) {
            accountManager.getAuthToken(account, getString(R.string.account_authentication_token_type), null, this, null, null);
            return true;
        } else {
            noAccountAdded();
            accountManager.addAccount(getString(R.string.account_type),getString(R.string.account_authentication_token_type),
                    null, null, this, null, null);
            return false;
        }
    }

    @Nullable
    @SuppressWarnings("MissingPermission")
    private Account getAccount() {
        AccountManager accountManager = AccountManager.get(this);
        Account[] accounts = accountManager.getAccountsByType(getString(R.string.account_type));
        if (accounts.length > 0) {
            return accounts[0];
        } else {
            return null;
        }
    }

    protected void noAccountAdded(){}
}

File: velolabs_repos/operator-android/app/src/main/java/io/lattis/operator/presentation/base/activity/BaseAuthenticatedActivity.kt
package io.lattis.operator.presentation.base.activity

import android.accounts.Account
import android.accounts.AccountManager
import io.lattis.operator.R
import io.lattis.operator.presentation.base.BaseView
import io.lattis.operator.presentation.ui.base.activity.BaseActivity

open abstract class BaseAuthenticatedActivity<Presenter : ActivityPresenter<V>,V: BaseView> :
    BaseActivity<Presenter, V>() {
    override fun onResume() {
        authenticateAccount()
        super.onResume()
    }

   fun authenticateAccount(): Boolean {
        val account = account
        val accountManager = AccountManager.get(this)
        return if (account != null) {
            accountManager.getAuthToken(
                account,
                getString(R.string.account_authentication_token_type),
                null,
                this,
                null,
                null
            )
            true
        } else {
            accountManager.addAccount(
                getString(R.string.account_type),
                getString(R.string.account_authentication_token_type),
                null,
                null,
                this,
                null,
                null
            )
            false
        }
    }

    private val account: Account?
        private get() {
            val accountManager = AccountManager.get(this)
            val accounts =
                accountManager.getAccountsByType(getString(R.string.account_type))
            return if (accounts.size > 0) {
                accounts[0]
            } else {
                null
            }
        }
}